### PR TITLE
RSDK-8663: Optimize away gostream concurrency complexity for camera clients.

### PR DESCRIPTION
**Pull request by dgottlieb**

### CODE REVIEW

**Comment by hexbabe** [comment]
@dgottlieb looks like CI is failing FYI

**Comment by dgottlieb** [comment]
Thanks -- did you otherwise review this?

**Review by hexbabe — COMMENTED** [review]
Did we pinpoint the reason why two GetImage calls are sometimes made? Is it possible to prevent that instead

**Comment by dgottlieb** [comment]
> Did we pinpoint the reason why two GetImage calls are sometimes made?

Yes. The ticket/comments try to explain what happens. But it is hard to follow how that relates to the exact code because of the gostream layers involved.

> Is it possible to prevent that instead

Given this patch does prevent that with camera clients, I assume you mean in the more general case? Yes and no. Right now, the API exposed by all [underlying gostream cameras](https://github.com/viamrobotics/rdk/blob/a302fe7e39fb4156c3a3063a9dd12217f86687e0/components/camera/camera.go?plain=1#L115-L119) are:
* `Images`
* `Stream`

Beyond those two, cameras may not have any* other methods[1]. So for those that do, we can try and guess/use them. But even for the `Read` [method](https://github.com/viamrobotics/rdk/blob/ebedca2ab41890d9ab8ba053a50efc0d80b96d79/components/camera/client.go#L121) that camera clients have, I'm not seeing many other cameras implement that. For example, the [h264 camera](https://github.com/viam-modules/viamrtsp/blob/84cf920d47727ca8d4239d5804defee0e8ce1497/rtsp.go#L723) becomes a "[sourceBasedCamera](https://github.com/viamrobotics/rdk/blob/ebedca2ab41890d9ab8ba053a50efc0d80b96d79/components/camera/videosourcewrappers.go#L39-L45)" which only has the Images/Stream APIs + the rtp passthrough apis.

Point being, avoiding "double get image" from a bigger set of cameras requires more typing across the codebase. And the camera client was 90% of our problem today.

[1] As far as I can tell -- it's _not_ correct to blindly use `[Get]Images(req)[0]` as a substitute for `[Get]Image`, e.g:
```
func (s *Server) GetImage(req *ImageRequest) {
  return camera.Images(...)[0]
}
```

**Comment by hexbabe** [comment]
Thanks for that info

From the writeup on the ticket:
> The issue is, the background thread in camera client's Stream would have already called the c.Read() again, so the second request will be invoked and almost immediately cancelled. 

I guess my question was more pointed towards this. There isn't a clear path forward preventing the double call originating from this background thread is there?

File: components/camera/server_test.go [hunk 1]
```diff
@@ -506,3 +506,17 @@ func TestServer(t *testing.T) {
...
+
+//nolint
+func TestCameraClientIsReadImager(t *testing.T) {
+	// RSDK-8663: Enforce that a camera client always satisfies the optimized `ReadImager`
+	// interface.
+	cameraClient, err := camera.NewClientFromConn(nil, nil, "", resource.Name{}, nil)
+	test.That(t, err, test.ShouldBeNil)
+
+	if _, isReadImager := cameraClient.(camera.ReadImager); isReadImager {
+		// Success
```
[the remark this answers is not in the document]
Comment (dgottlieb) [reply components/camera/server_test.go:518]: How this code got this way is a bit of a long story -- it wasn't due to preference. I was trying to use the "test.ShouldImplement" assertion, that I think requires reflection and additional error handling. But ultimately failed at that. So the structure of this if-statement was already there and I just had to use a type-assertion instead and do nothing in the "then" clause.

That said, I do think writing conditionals in the affirmative (i.e: avoid double negatives) reads better. And in this case, a real writer is going to write the if-statement that way -- just with a non-empty "then".

But normally I wouldn't take that to the "extreme" of having an empty "then" clause. I just left it this way because I was fed up with writing an assertion.

File: components/camera/client.go [hunk 2]
```diff
@@ -118,6 +118,8 @@ func getExtra(ctx context.Context) (*structpb.Struct, error) {
 	return ext, nil
 }
 
+// RSDK-8663: This method signature is depended on by the `camera.serviceServer` optimization that
+// avoids using an image stream just to get a single image.
```
[the remark this answers is not in the document]
Comment (dgottlieb) [reply components/camera/client.go:122]: That's exactly how this patch works yes. Instead of spinning up:
* stream = camera.Stream()
* stream.Next()
* stream.Close

We first ask "Hey camera, do you have a "Read" method?" Great -- we'll just call that one time and avoid this stream non-sense.

**Comment by dgottlieb** [comment]
> I guess my question was more pointed towards this. There isn't a clear path forward preventing the double call originating from this background thread is there?

I explored an avenue where ReadImage/ReadMedia would do:
* stream = camera.Stream()
* image = stream.NextAndClose()

Which I think is what you're asking? 

I didn't find that to be possible without significantly mucking with the producerConsumer code. Which would then require a much more thorough proof of correctness/review.

**Comment by hexbabe** [comment]
> > I guess my question was more pointed towards this. There isn't a clear path forward preventing the double call originating from this background thread is there?
> 
> I explored an avenue where ReadImage/ReadMedia would do:
> 
> * stream = camera.Stream()
> * image = stream.NextAndClose()
> 
> Which I think is what you're asking? Where we can "atomically" get a single image and communicate to the "producer" to not (eagerly) produce the next image.
> 
> I didn't find that to be possible without significantly mucking with the producerConsumer code. Which would then require a much more thorough proof of correctness/review.

Makes sense. That sounds very involved

File: components/camera/server_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (hexbabe) [reply components/camera/server_test.go:518]: Haha the test lib use of reflection can be super funky

Got it
